### ISSUE TITLE
[meta] Ability to add search engines and or remove existing search engines

### ISSUE THREAD

**Issue by bbinto** [mozilla-mobile/focus-android#961]
### MVP
- [x] #1484 Update search engines to use radio buttons
- [x] #1556  String update
- [x] #1560 Fix UI Tests
- [x] #1617 Add screen for manual creation of search engines (behind feature flag)
- [x] #1660 Persist search engines (behind feature flag)
- [x] #1671 Simplify Settings screens with enums
- [x] #1644 Allow removing of search engines (behind feature flag)
- [x] #1618 Engine radio buttons style polish
- [x] #1689 Search engine polish, Screenshots UITest [seeing UI Test failures, will land on return]
- [x] #1645 Pref on for 3.0
- [x] #1647 search engine ordering
- [x] #1691 sub text in edit screen
- [x] #1705 change example text
- [ ] #1828 Add SUMO page for search engine
- [x] #1755 [ux] review
- [x] #1755 UX implementation feedback
- [x] #1856 Add telemetry

(#1498 reference to closed breakdown)

----
#### I have difficulties finding a way to change my default search engine
----
Mocks
![flow - add search engine manually](https://user-images.githubusercontent.com/287322/32025648-f30183e0-b995-11e7-852f-6869e05b85d2.png)
![flow - remove restore search engine](https://user-images.githubusercontent.com/287322/32025649-f31aaeba-b995-11e7-8ea7-183f70606a0e.png)

![visual design](https://user-images.githubusercontent.com/287322/32027376-8c9af278-b99d-11e7-8999-e6696dcb347b.png)

**Comment by pocmo** [comment]
See also #831 and #808

**Comment by bbinto** [comment]
Talked to legal, this should be fine to do

**Comment by bbinto** [comment]
More feature requests: 
- https://play.google.com/store/apps/details?id=org.mozilla.klar&reviewId=Z3A6QU9xcFRPRjN3Q1VKcWlhQmphQWQ1R2NnSVdRNVVxWVpqdUZHVVVtTUJuNi1HMWQ1Mk9KUTNKZ2ZsMzBKLTdzYTE2UXM0bjM0cVAta2kyOHNtOUxIb3ow

**Comment by antlam** [comment]
@bbinto I'm going to move this to Granite* since Folly is a non-UX release and already has "What's New?"

Let me know if you disagree! happy to discuss more when you're back 

Also, the same feature is slated for Granite* for iOS! so that would be kind of nice to match :)

**Comment by BrianNJones** [comment]
As discussed, revamping Settings will require significant level of UX/content effort. Recommendation to move to Indigo. Also impacts #1054.

**Comment by lime124** [comment]
@BrianNJones commented on the wrong bug...?

We can get this search effort in and are in progress on the designs. As discussed with @Sdaswani, we'll be able to finalized strings in by end of day Tuesday to pass of to Eng for export.

**Comment by pocmo** [comment]
Hopefully some helpful input:

On Fennec there are two ways how the user can add a search engine:
* Long-pressing an input field and selecting "Add search engine". The big advantage is that this works for a variety of sites. Can be confusing for fields that are not actually a search field and sometimes this might not work well. We'd have to investigate if we can implement this with WebView
* The second option is to use the menu: page -> add search engine. This works for all websites that define in their metadata that they are a search and how to perform a search. The nice thing is that we get everything from the provider (name, URL, etc.) and therefore it's always working. This mechanism even supports more features like search suggestions (but we are not using this in Focus). Of course this does not work with every page. But every major search engine does support it.

Constraints:
* The user has different search engines depending on his language.
* The user can switch the language and this changes the search engines that are available.

**Comment by pocmo** [comment]
Oh, and there's a third option in Fennec. You can add them via AMO, e.g.:
https://addons.mozilla.org/en-US/android/addon/metager-metasuche/

**Comment by antlam** [comment]
[Searchengine-add.zip](https://github.com/mozilla-mobile/focus-android/files/1380508/Searchengine-add.zip)

Icons!

**Comment by liuche** [comment]
Landed strings in #1453.

**Comment by liuche** [comment]
For reference, since we use `AppCompatActivity`, nested `PreferenceScreen` will lack the ActionBar/Toolbar because the SupportActionBar and tho ActionBar conflict ([ref](https://stackoverflow.com/questions/29357292/keep-the-actionbar-displayed-in-when-changing-preferencescreen/36116481#36116481)). We also don't use [PreferenceFragmentCompat](https://developer.android.com/reference/android/support/v7/preference/PreferenceFragmentCompat.html) probably because it came out in r23, though it's something we can look into at some point (but it also sounds like it's got some bugs).

Adding a comment in-code too, to be clearer why we can't just use nested `PreferenceScreen`.

**Comment by pocmo** [comment]
> We also don't use PreferenceFragmentCompat probably because it came out in r23, though it's something we can look into at some point (but it also sounds like it's got some bugs).

At some point I tried to migrate to the PreferenceFragment from the support library. I wanted to get rid of the additional activity and look into whether we can just have the preferences as another fragment that gets added to our MainActivity. This turned out to be more complicated than I thought. The APIs were partially different and our custom preferences didn't work either and were hard to port. In the end I stopped and moved on.

**Comment by liuche** [comment]
WIP progress here: https://github.com/liuche/focus-android/tree/searchengine

I think that I've got the complicated parts out of the way (understanding how our Settings Fragments work, adding new screens, showing search engines as radio buttons), and to finish the very minimum of "add a search engine", I just need to add the "Manually Add Search Engine" (which is probably a Small-Medium).

Beyond that, this will still take more work to add the other screens: tutorial, as well as some exploration regarding the "automatically add a search engine", and removing/restoring default search engines.

I would say "automatically extract and add the right search engine information" is a Large, the Tutorial is a Small-Medium (but depends on which approach we take for "automatically add"), and removing/restoring default search engines is a Medium.

@bbinto, what are your thoughts on shipping only "Manually add search engines" with this release? (It has some minimal explanation in the UI, but depending on what the user types in, could also fail with incorrect formatted input.)

Or do we need to also include 1) automatically adding the right search engine urls and name if a user chooses to add a search engine from a website (rather than manually typing in the search engine query url), 2) tutorial, and 3) remove/restore (default) search engines?

**Comment by Sdaswani** [comment]
I think given the new schedule, this doesn't need to be split up into two releases. Put it behind a feature flag and land what you have now?

**Comment by bbinto** [comment]
To me it's not ideal to just have "manually add search engine" in the next release. However, if we decide to go with that, we *need* to offer a way to remove the entry for sure in case the user typed in the wrong URL, or at least a way to change an invalid search engine entry.

**Comment by lime124** [comment]
![flow - add search engine manually](https://user-images.githubusercontent.com/287322/32025648-f30183e0-b995-11e7-852f-6869e05b85d2.png)
![flow - remove restore search engine](https://user-images.githubusercontent.com/287322/32025649-f31aaeba-b995-11e7-8ea7-183f70606a0e.png)

**Comment by lime124** [comment]
Made this simpler for this first version and created a ticket to iterate on it. 
https://github.com/mozilla-mobile/focus-android/issues/1561

**Comment by TPS** [comment]
Since this has to remain privacy-focused, I'd like to RFC/E the following related to this issue _w/o (yet)_ creating a full issue:

There's a lot of blank _(useless)_ space on the opening screen.  IMHO, an _optional_ list of installed search engine fields may answer to use that space constructively.  Any interest?

**Comment by lime124** [comment]
Hey @TPS - if I understand your proposal correctly, you are suggesting we list installed search engines on the homescreen (see image)? Can you summarize the problem this proposal solves? Thanks!

![homescreen - keyboard - toast](https://user-images.githubusercontent.com/287322/32563605-84e2d91e-c466-11e7-9f1f-6ebdc4440d1b.png)

**Comment by TPS** [comment]
I'll expound a little:   All that blank space above could serve us better.  In regular Ff, we use tiles w/ user history.  Since Focus/Klar doesn't have that, maybe search fields/tiles could optionally serve the same purpose of constructively using the space.

**Comment by lime124** [comment]
Thanks for the follow up @TPS. Could you open up a new ticket for us to track that idea? Thanks!

**Comment by TPS** [comment]
@lime124 I've opened #1692, per your encouragement & suggestion.  I'd like to delete my end of the discussion here (now duplicated there) to cleanup this meta-issue.  Any objections?

**Comment by lime124** [comment]
go for it! thanks @TPS!

**Comment by liuche** [comment]
@lime124 I landed the last piece of this, so when everything goes green you should get a build and can do a final UX review!

**Comment by liuche** [comment]
Unassigning myself because all eng work is done.
